### DISCUSSION TITLE
Update the session context to the decorator？

### DISCUSSION THREAD

**Discussion by wu-clan** [fastapi-practices/fastapi_best_architecture#19] [Ideas]
It feels good.

```python
from functools import wraps

def session_decorator(func):
    @wraps(func)
    async def wrapper(*args, **kwargs):
        async with async_db_session.begin() as db:
            return await func(db, *args, **kwargs)
    return wrapper
```

If use it, need to change a more practical method name.

**Comment by wu-clan** [comment]
Need to test

- [x] services
- [x] api (When used with the request method decorator)

**Comment by wu-clan** [comment]
Hi, @downdawn 

What do you think?

**Comment by wu-clan** [comment]
Not all db operations require transactions

update `async_db_session` to `async_session_maker`

E.g.:
```python
def async_session(func):
    """
    session 装饰器

    :param func:
    :return:
    """

    async def wrapper(*args, **kwargs):
        async with async_session_maker() as db:
            return await func(db, *args, **kwargs)

    return wrapper


def async_session_begin(func):
    """
    session 事务装饰器

    :param func:
    :return:
    """

    async def wrapper(*args, **kwargs):
        async with async_session_maker.begin() as db:
            return await func(db, *args, **kwargs)

    return wrapper
```

**Reply by downdawn** [reply]
Agree

async_session_begin or async_session_transaction

**Comment by wu-clan** [comment]
Suspend the merger #20

When using this decorator, the IDE will lose parameter prompts and checks.

**Reply by downdawn** [reply]
Is there a screenshot?

**Reply by wu-clan** [reply]
No, You can check in pycharm

When using this decorator, in api functions, the caller will not be prompted for missing arguments, and there will be no error alert when using `*` mandatory formal parameters

**Reply by downdawn** [reply]
I see

This may be unsolvable, because the decorator cannot know which parameters the decorated function needs, and can only check whether these parameters exist by passing the parameter name.

**Reply by wu-clan** [reply]
Yes, this affects developers to some extent to do error checking

**Reply by downdawn** [reply]
May have to make a trade off

**Comment by downdawn** [comment]
Hi, @wu-clan 

Why not just use：`CurrentSession = Annotated[AsyncSession, Depends(get_db)]`

And multi-table query will cause exception
`sqlalchemy.exc.StatementError: (sqlalchemy.exc.MissingGreenlet) greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place?`

like the userinfo interface

**Reply by wu-clan** [reply]
Do not use `CurrentSession = Annotated[AsyncSession, Depends(get_db)]` globally because this is only used as an alternate option, choose according to your preference, you can see examples in `jwt.py`, otherwise it will add a db parameter to all api functions

**Reply by wu-clan** [reply]
Multi-table queries and single-table queries will be different in async, just using CurrentSession can solve.

When async, you must ensure that multiple table queries use the same session

**Reply by wu-clan** [reply]
This was an early question I had, but I can't remember the details

https://github.com/sqlalchemy/sqlalchemy/discussions/7835

**Reply by wu-clan** [reply]
Perhaps this needs attention as well

https://github.com/sqlalchemy/sqlalchemy/discussions/8176

**Reply by downdawn** [reply]
Help to evaluate the feasibility of this

```
async def get_db() -> AsyncSession:
    """
    session 生成器
    """
    async with async_db_session() as session:
        try:
            yield session
        except Exception as se:
            await session.rollback()
            raise se
        finally:
            await session.close()


async def get_db_transaction() -> AsyncSession:
    """
    session 生成器 事务
    """
    async with async_db_session.begin() as session:
        try:
            yield session
        except Exception as se:
            await session.rollback()
            raise se
        finally:
            await session.close()


# Session 依赖注入
CurrentSession = Annotated[AsyncSession, Depends(get_db)]
CurrentSessionTransaction = Annotated[AsyncSession, Depends(get_db_transaction)]
```

**Reply by wu-clan** [reply]
`async with async_db_session() as session:` is typically used for queries without caring about transaction rollbacks, and it automatically closes the session
`async with async_db_session.begin() as session:` already contains the transaction and does not need to be manually processed

I think they all just need a `yield session`

**Comment by wu-clan** [comment]
This solution has been abandoned.
